src/core/namespace.c

```diff
@@ -3181,6 +3183,117 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+int refresh_extensions_in_namespace(
+                const PidRef *target,
+                const char *hierarchy_env,
+                const NamespaceParameters *p) {
+
+        const char *overlay_prefix = "/run/systemd/mount-rootfs";
+        _cleanup_(mount_list_done) MountList ml = {};
+        _cleanup_free_ char *extension_dir = NULL;
+        _cleanup_strv_free_ char **hierarchies = NULL;
+        MountInNamespaceFlags min_flags = 0;
+        int r;
+
+        assert(pidref_is_set(target));
+        assert(hierarchy_env);
+        assert(p);
+
+        log_debug("Refreshing extensions in-namespace for hierarchy '%s'", hierarchy_env);
+
+        extension_dir = path_join(p->private_namespace_dir, "unit-extensions");
+        if (!extension_dir)
+                return -ENOMEM;
+
+        min_flags |= MOUNT_IN_NAMESPACE_MAKE_FILE_OR_DIRECTORY;
+
+        r = parse_env_extension_hierarchies(&hierarchies, hierarchy_env);
+        if (r < 0)
+                return r;
+
+        r = append_extensions(
+                        &ml,
+                        overlay_prefix,
+                        p->private_namespace_dir,
+                        hierarchies,
+                        p->extension_images,
+                        p->n_extension_images,
+                        p->extension_directories);
+        if (r < 0)
+                return r;
+        if (ml.n_mounts == 0)
+                return 0;
```

Comment:
hmm, is this right? if i apply some confexts first, and then remove them all, we should unmount all confexts again

Reply:
also, we need to sort the mounts here, i.e. call drop_unused_mounts() (which should probably get a better name, maybe sort_and_drop_unused_mounts())

Reply:
> hmm, is this right? if i apply some confexts first, and then remove them all, we should unmount all confexts again

The overlay is always added to the list, even when there is only the "root" filesystem in it, with no extensions. If this returns 0, then it means there's really nothing to do

Reply:
> also, we need to sort the mounts here, i.e. call drop_unused_mounts() (which should probably get a better name, maybe sort_and_drop_unused_mounts())

Wouldn't hurt I guess, but would it be useful either? This is just adding the overlay with the /etc/ hierarchy, there's not going to be any overlap or ordering or anything else involved? IE, there are no bind mounts

Reply:
> The overlay is always added to the list, even when there is only the "root" filesystem in it, with no extensions. If this returns 0, then it means there's really nothing to do

Again: what happens if i have 5 confext applied, then delete them all and reload? We need to unmount them in that case, not leaving them in place. as i understand this code here you are shortcutting the whole operation completely, levaing the 5 confext in place. that's plain wrong!

Reply:
> Wouldn't hurt I guess, but would it be useful either? This is just adding the overlay with the /etc/ hierarchy, there's not going to be any overlap or ordering or anything else involved? IE, there are no bind mounts

Please dont cut corners.

Reply:
> > Wouldn't hurt I guess, but would it be useful either? This is just adding the overlay with the /etc/ hierarchy, there's not going to be any overlap or ordering or anything else involved? IE, there are no bind mounts
> 
> Please dont cut corners.

It's not cutting corners? I'm just asking if it would make any practical difference, or if it would be a no-op. It can be added even if it's a no-op, it's fine, just trying to understand

Reply:
> > The overlay is always added to the list, even when there is only the "root" filesystem in it, with no extensions. If this returns 0, then it means there's really nothing to do
> 
> Again: what happens if i have 5 confext applied, then delete them all and reload? We need to unmount them in that case, not leaving them in place. as i understand this code here you are shortcutting the whole operation completely, levaing the 5 confext in place. that's plain wrong!

They are removed, that's what I am saying, it works as you want it to. I even showed that to you at ASG in the demo, go check the recording if you don't remember ;-)
The reason it works is that as mentioned is that n_mounts is not zero, even when there are no extensions. There's always the overlay. So if this is 0 it means that there's really, really nothing to do or undo.

Reply:
so, sorting things shouldn't make the big difference, you should be adding your images in sorted order anyway. but who knows, if people actually use the hierarchy override env var, then things might get weird. hence, please apply the correct order, so that we are strictly systematic, and things don't depend on the order of user configuration.